File: Awful.apk/src/main/java/com/ferg/awfulapp/task/ReportRequest.java
package com.ferg.awfulapp.task;

import android.content.Context;
import android.net.Uri;

import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.util.AwfulError;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

/**
 * Created by matt on 8/8/13.
 */
public class ReportRequest extends AwfulRequest<String> {
    private int postId;
    private String mComments;
    public ReportRequest(Context context, int postId, String mComments) {
        super(context, null);
        this.mComments = mComments;
        this.postId = postId;
    }

    @Override
    protected String generateUrl(Uri.Builder urlBuilder) {
        addPostParam(Constants.PARAM_COMMENTS, mComments);
        addPostParam(Constants.PARAM_POST_ID, Integer.toString(postId));
        addPostParam(Constants.PARAM_ACTION, "submit");

        return Constants.FUNCTION_REPORT;
    }

    @Override
    protected String handleResponse(Document doc) throws AwfulError {

		if(doc.getElementById("content") != null){
			Element standard = doc.getElementsByClass("standard").first();
			if(standard != null && standard.hasText()){
				if(standard.text().contains("Thank you, but this thread has already been reported recently!")){
					throw new AwfulError("Someone has already reported this thread recently");
					
				}else if(standard.text().contains("Your alert has been submitted to the Moderators.")){
					return "Your alert has been submitted to the Moderators."; //"Thank you for your report";
				}
			}
			throw new AwfulError("An error occurred while trying to process your report");
		}
		throw new AwfulError("An error occurred while trying to send your report");
    }

    @Override
    protected boolean handleError(AwfulError error, Document doc) {
        return error.isCritical();
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/task/ReportRequest.kt
package com.ferg.awfulapp.task

import android.content.Context
import com.ferg.awfulapp.constants.Constants.*
import com.ferg.awfulapp.util.AwfulError
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element

/**
 * A request that submits a report for a post.
 *
 * This sends a report to the mods for a given [postId], providing [comments] as
 * the report reason. The response is a status message when the report was
 * successful, otherwise an error is thrown.
 */
class ReportRequest(context: Context, private val postId: Int, private val comments: String)
    : AwfulRequest<String>(context, FUNCTION_REPORT, isPostRequest = true) {

        init {
        with(parameters) {
            add(PARAM_COMMENTS, comments)
            add(PARAM_POST_ID, postId.toString())
            add(PARAM_ACTION, "submit")
        }
    }

    @Throws(AwfulError::class)
    override fun handleResponse(doc: Document): String {
        val message = doc.selectFirst("standard")?.takeIf(Element::hasText)?.text()
        fun responseSays(text: String) = message?.contains(text, ignoreCase = true) ?: false
        // TODO: when I tested this I got a toast with a different success message - no clue where it came from!
        return when {
            responseSays("your alert has been submitted") ->
               "Report sent!"
            responseSays("this thread has already been reported") ->
                "Someone has already reported this thread recently"
            else ->
                throw AwfulError("An error occurred while trying to send your report")
        }
    }

}
